stable_baselines3/her/her_replay_buffer.py

```diff
@@ -185,347 +116,235 @@ def set_env(self, env: VecEnv) -> None:
+            # so the next achieved_goal depends also on the previous state and action
+            # because we are in a GoalEnv:
+            # r_t = reward(s_t, a_t) = reward(next_achieved_goal, desired_goal)
+            # therefore we have to use next_obs["achieved_goal"] and not obs["achieved_goal"]
+            next_obs["achieved_goal"],
+            infos,
+            # we use the method of the first environment assuming that all environments are identical.
+            indices=[0],
+        )
+        rewards = rewards[0].astype(np.float32)  # env_method returns a list containing one element
+        obs = self._normalize_obs(obs, env)
+        next_obs = self._normalize_obs(next_obs, env)
+
+        # Convert to torch tensor
+        observations = {key: self.to_torch(obs) for key, obs in obs.items()}
+        next_observations = {key: self.to_torch(obs) for key, obs in next_obs.items()}
+
+        return DictReplayBufferSamples(
+            observations=observations,
+            actions=self.to_torch(self.actions[batch_inds, env_indices]),
+            next_observations=next_observations,
+            # Only use dones that are not due to timeouts
+            # deactivated by default (timeouts is initialized as an array of False)
+            dones=self.to_torch(self.dones[batch_inds, env_indices] * (1 - self.timeouts[batch_inds, env_indices])).reshape(
+                -1, 1
+            ),
+            rewards=self.to_torch(self._normalize_reward(rewards.reshape(-1, 1), env)),
+        )
+
+    def _sample_goals(self, batch_inds: np.ndarray, env_indices: np.ndarray) -> np.ndarray:
         """
         Sample goals based on goal_selection_strategy.
-        This is a vectorized (fast) version.
 
-        :param episode_indices: Episode indices to use.
-        :param her_indices: HER indices.
-        :param transitions_indices: Transition indices to use.
-        :return: Return sampled goals.
+        :param trans_coord: Coordinates of the transistions within the buffer
+        :return: Return sampled goals
         """
-        her_episode_indices = episode_indices[her_indices]
+        batch_ep_start = self.ep_start[batch_inds, env_indices]
+        batch_ep_length = self.ep_length[batch_inds, env_indices]
 
         if self.goal_selection_strategy == GoalSelectionStrategy.FINAL:
             # replay with final state of current episode
-            transitions_indices = self.episode_lengths[her_episode_indices] - 1
+            transition_indices_in_episode = batch_ep_length - 1
 
         elif self.goal_selection_strategy == GoalSelectionStrategy.FUTURE:
             # replay with random state which comes from the same episode and was observed after current transition
-            transitions_indices = np.random.randint(
-                transitions_indices[her_indices], self.episode_lengths[her_episode_indices]
-            )
+            current_indices_in_episode = batch_inds - batch_ep_start
```

Comment:
does that work also when the episode is split in the buffer?

Reply:
oh i see, you do the wrapping later but I still have a doubt because of `np.random.randint(current_indices_in_episode, batch_ep_length)`

Reply:
Yes, it is handled here:

```python
transition_indices = (transition_indices_in_episode + batch_ep_start) % self.buffer_size
```

I also have a doubt for FUTURE when episode is split. I changed it. If it worked before, it was a fluke.